src/token-service/metadata-token-service.ts

```diff
@@ -15,64 +21,93 @@ export class MetadataTokenService implements TokenService {
     private readonly url: string;
     private readonly opts: Options;
     private token?: string;
+    private tokenExpiredAt = 0;
+    private tokenTimeToRefresh = 0;
+    private currentFetchToken?: Promise<string>;
```

Comment:
I would suggest to use something like https://github.com/DirtyHairy/async-mutex

Reply:
I looked at this library. It is about mutexes and simophores, which doesn't seem solving my task. I need that one internal query is formed for several parallel queries, and all queries return the same result - promis sharing solves it